src/symbol.rs

```diff
@@ -0,0 +1,496 @@
+use std::fmt;
+
+use crate::parser::LURK_WHITESPACE;
+#[cfg(not(target_arch = "wasm32"))]
+use proptest_derive::Arbitrary;
+/// Module for symbol type, Sym.
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+pub const KEYWORD_MARKER: char = ':';
+pub const SYM_SEPARATOR: char = '.';
+pub const SYM_MARKER: char = '.';
+pub const ESCAPE_CHARS: &str = "|(){}[],.:'\\\"";
+
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Serialize, Deserialize, Hash)]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Arbitrary))]
+/// Type for hierarchical symbol names
+pub struct Symbol {
+    pub path: Vec<String>,
+}
+
+impl Symbol {
+    /// Creates a new Symbol::Sym with the root path `[]`.
+    #[inline]
+    pub fn root() -> Self {
+        Self { path: vec![] }
+    }
+
+    /// Returns true if the Symbol is the root sym or keyword, i.e. if it has a path of `[]`.
+    pub fn is_root(&self) -> bool {
+        self.path.is_empty()
+    }
+
+    /// Creates a new Symbol with the path extended by the given vector of path segments.
+    pub fn new<A: AsRef<str>>(path: &[A]) -> Self {
+        Self {
+            path: path.iter().map(|x| String::from(x.as_ref())).collect(),
+        }
+    }
+
+    /// Creates a new Symbol with the path extended by the given vector of path segments.
+    pub fn extend<A: AsRef<str>>(&self, child: &[A]) -> Self {
+        let mut path = Vec::with_capacity(self.path.len() + child.len());
+        for elt in self.path.iter() {
+            path.push(elt.clone());
+        }
+        for elt in child.iter() {
+            path.push(String::from(elt.as_ref()));
+        }
+        Self { path }
+    }
+
+    pub fn nil() -> Symbol {
+        Self::new(&["lurk", "nil"])
+    }
+
+    pub fn keyword<A: AsRef<str>>(key: &[A]) -> Symbol {
+        Self::new(&["keyword"]).extend(key)
+    }
```

Comment:
Can we have a symbol (not a keyword) whose path is `["keyword".to_string()]`? Why use the same underlying data layout to represent both symbols and keywords?